langfuse/callback/llama_index.py

```diff
@@ -0,0 +1,369 @@
+)
+from langfuse.decorators.error_logging import (
+    auto_decorate_methods_with,
+    catch_and_log_errors,
+)
+from langfuse.callback.base import BaseCallbackHandler as LangfuseBaseCallbackHandler
+from langfuse.callback.utils import CallbackEvent
+
+try:
+    from llama_index.core.callbacks.base_handler import (
+        BaseCallbackHandler as LLamaIndexBaseCallbackHandler,
+    )
+    from llama_index.core.callbacks.schema import (
+        CBEventType,
+        BASE_TRACE_EVENT,
+        EventPayload,
+    )
+    from llama_index.core.utilities.token_counting import TokenCounter
+except ImportError:
+    raise ModuleNotFoundError(
+        "Please install llama-index to use the Langfuse llama-index integration: 'pip install llama-index'"
+    )
+
+
+@auto_decorate_methods_with(catch_and_log_errors, exclude=["__init__"])
+class LLamaIndexCallbackHandler(
+    LLamaIndexBaseCallbackHandler, LangfuseBaseCallbackHandler
+):
+    log = logging.getLogger("langfuse")
+
+    def __init__(
+        self,
+        *,
+        public_key: Optional[str] = None,
+        secret_key: Optional[str] = None,
+        host: Optional[str] = None,
+        debug: bool = False,
+        stateful_client: Optional[
+            Union[StatefulTraceClient, StatefulSpanClient]
+        ] = None,
+        session_id: Optional[str] = None,
+        user_id: Optional[str] = None,
+        trace_name: Optional[str] = None,
+        release: Optional[str] = None,
+        version: Optional[str] = None,
+        threads: Optional[int] = None,
+        flush_at: Optional[int] = None,
+        flush_interval: Optional[int] = None,
+        max_retries: Optional[int] = None,
+        timeout: Optional[int] = None,
+        event_starts_to_ignore: Optional[List[CBEventType]] = None,
+        event_ends_to_ignore: Optional[List[CBEventType]] = None,
+        tokenizer: Optional[Callable[[str], list]] = None,
+    ) -> None:
+        LLamaIndexBaseCallbackHandler.__init__(
+            self,
+            event_starts_to_ignore=event_starts_to_ignore or [],
+            event_ends_to_ignore=event_ends_to_ignore or [],
+        )
+        LangfuseBaseCallbackHandler.__init__(
+            self,
+            public_key=public_key,
+            secret_key=secret_key,
+            host=host,
+            debug=debug,
+            stateful_client=stateful_client,
+            session_id=session_id,
+            user_id=user_id,
+            trace_name=trace_name,
+            release=release,
+            version=version,
+            threads=threads,
+            flush_at=flush_at,
+            flush_interval=flush_interval,
+            max_retries=max_retries,
+            timeout=timeout,
+            sdk_integration="llama-index",
+        )
+
+        self.root = stateful_client
+        self.event_map: Dict[str, List[CallbackEvent]] = defaultdict(list)
+        self._cur_trace_id: Optional[str] = None
+        self._token_counter = TokenCounter(tokenizer)
+
```

Comment:
I think it would be great to have variable declaraitons for these at the top at class level to help our linters.